### ISSUE TITLE
be interactive

### ISSUE THREAD

**Issue by mafintosh** [juliangruber/electron-stream#4]
would be nice if this could execute the js as you stream it to it instead of buffering the entire source.
it could probably be executed like this

``` js
try {
  eval(sourceBuffer)
} catch (e) {
  if (e.name === 'SyntaxError') {
    // bufferMore
    return
  } else {
    throw err 
  }
}
sourceBuffer = ''
```

**Comment by thlorenz** [comment]
For that  [snippetify](https://github.com/thlorenz/snippetify) may help which will try to detect executable JS snippets in a given string.

**Comment by juliangruber** [comment]
great suggestion! i'm assuming snippetify will help with raw streaming data as chunk boundaries likely won't line up with line endings. i'll give it a shot!

**Comment by mafintosh** [comment]
@thlorenz @juliangruber just used snippetify to make a streaming snippet parser, https://github.com/mafintosh/snippet-stream

then its as simple as

```
var snippets = require('snippet-stream')

// create a snippet stream
var stream = snippets()
var scope = {}

// write some js to it
stream.write('var a = 1\n')
stream.write('function foo () {\n')
stream.write(' return a + 1\n')
stream.write('}\n')
stream.write('console.log(foo())')
stream.end()

stream.on('data', function (data) {
  eval.call(scope, data)
})
```

**Comment by juliangruber** [comment]
@mafintosh ok that's really sweet. Ok I might need your guys' help again:

In the current design I boot an http server that serves the script inlined into a <script> tag: https://github.com/juliangruber/electron-stream/blob/master/index.js#L77. I'm doing this because otherwise window.onerror will only get "Script Error." instead of a full error with stack trace.

Could you think of a way to serve javascript via a snippet stream, like using electron's ipc module, while still retaining sourcemap and full window.onerror support?

**Comment by RangerMauve** [comment]
If you use an HTTP server you can use the response stream and pipe data into it

```
res.write("<!DOCTYPE html><title>electron-stream</title><script>");
js.pipe(res);
js.on("end", function(){ res.end("</script>"});
```

IIRC the browser will evaluate the JS as it's piped into the script tag.

**Comment by juliangruber** [comment]
> IIRC the browser will evaluate the JS as it's piped into the script tag.

I did a test on this and unfortunately no it doesn't.

So maybe we introduce this as a second mode simply, you lose sourcemaps and window.onerror support but gain interactive execution.

**Comment by juliangruber** [comment]
IIUC some javascript will even break by this execution technique, imagine for example those two snippets arriving in separate chunks

```js
console.log(typeof foo)
```

```js
var foo = 'bar'
```

The output is going to be `undefined`, whereas in full execution it would be `"string"`, thanks to variable hoisting.

Therefore interactive mode is strictly to be seen as interactive / repl mode, it's not a way of improving performance for regular scripts.

**Comment by RangerMauve** [comment]
@juliangruber Just to confirm, in your example it wouldn't be "string". Variable hosting does not hoist the assignment. It would still be undefined until the line when the value gets assigned to the variable.

```javascript
console.log(typeof foo);
var foo = 'bar'
```

Turns into

```javascript
var foo
console.log(typeof foo)
foo = 'bar'
```

What it wouldn't work with is function hoisting since the value does get hoisted to the top in that case.

**Comment by juliangruber** [comment]
yeah you're right about that, thanks for the clarification. Anyway, the important part is that interactive mode can break javascript if not done correctly.
